[Woo POS] Prevent foreground notices from showing in POS mode

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13635 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR prevents the "You have a new order" `Notices` from showing while in POS mode, while keeping the same functionality in App mode.

Since these notices are created in response to a push notification, and we would need some sort of backend support to adjust these better, the simplest solution for the time being seems to be to allow the push notification to go through anyway (so the user could still tap on their phone if the app is in background, and be redirected to the store's order) but hide them from the app's foreground if the POS mode is active.

## Changes
It's the first time I work with push notifications, so feel free to suggest any better way to handle this:
- We leverage the existing mechanism we use between app and pos to hide/show the tab bars to also enable/disable in-app notices, by passing a boolean to the `AppDelegate` to let the system now we're in POS mode.
- This will toggle `inAppNotices` true/false for the whole app instance
- `inAppNotices` is used as an additional check when deciding if we should present them to the merchant, in case that POS mode is active, then we return early with `UNNotificationPresentationOptions(rawValue: 0)`

## Testing
- In POS mode, go through the order processing flow and observe how at the end there's no longer any "You have a new order" notice.
- Repeat the same process in App, go through the order processing flow and observe how you see the usual "You have a new order" notice at the end of the flow.

Apologies for the iPhone testing video, my cable just died and I had to bypass the eligibility checker to use the alternative 😅, the end result should be the same for ipad but I could only test in iPhone. Worth noting as well that we can only test in physical devices, as push notifications don't work on simulator.

While developing I found that sometimes we do not receive push notifications, but this seems to be an existing issue as happens in trunk as well, but is worth to try more than once just to be sure is not a false positive.

From my testing:
- In POS mode, no in-app notices are shown.
- In App mode, in-app notices are shown for app orders.
- When moving to background from POS mode, background notifications still work as expected.
- When moving to foreground to POS mode, no in-app notices are shown.

https://github.com/user-attachments/assets/95589d33-9bc5-43cb-8937-26e4789c73bf

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Intercept in-app notices if app in POS mode

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/219">20.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Return early if `inAppNotices` is false

If we perform this check once we’ve already created the local notification, it seems to be ignored in some instances and it’s sent to the Publisher. We can avoid this by just returning earlier with `UNNotificationPresentationOptions(rawValue: 0)`

iamgabrielma committed:
Rename variables. Add comments

iamgabrielma committed:
linting

iamgabrielma committed:
Update mock and unit tests. make test target compile

joshheald commented:
> Apologies for the iPhone testing video, my cable just died and I had to bypass the eligibility checker to use the alternative 😅, the end result should be the same for ipad:

It's not even _that_ bad in POS with the phone in portrait. Well, it is, but it feels like there's a route through. I wonder if we had the product list and cart stacked vertically, with horizontal scrolling for each, how that would feel... maybe weird, but weird can be good...

joshheald commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -274,12 +274,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 //
 extension AppDelegate {
-    func setShouldHideTabBar(_ hidden: Bool) {
+    func updateSharedConfiguration(_ isPointOfSaleEnabled: Bool) {
```
```suggestion
    func updateSharedConfiguration(_ isPointOfSaleActive: Bool) {
```

iamgabrielma replied:
Updated https://github.com/woocommerce/woocommerce-ios/pull/13670/commits/d75f81f00b281dc596d6b725bcb8d12b7d99c910

joshheald commented on WooCommerce/Classes/AppDelegate.swift:
```diff
@@ -274,12 +274,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-    func setShouldHideTabBar(_ hidden: Bool) {
+    func updateSharedConfiguration(_ isPointOfSaleEnabled: Bool) {
+        // Show/hide app's tab bars
         guard let tabBarController = AppDelegate.shared.tabBarController else {
             return
         }
```
I think we'd have problems if there was no `tabBarController`, but strictly there's no reason that this should prevent us from disabling the in app notifications – perhaps this should become an `if let`, or separated into another function.

iamgabrielma replied:
True that. The logic to enable/disable notifications could happen independently from the `tabBarController` guard. I thought of creating 2 different functions but ultimately put everything in one. 

I think we can refactor this bit further since shouldn't be any more items (I believe?) that we share between app and pos but that must behave differently, and also add some DDLogs if the `guard` or `if let` fail (they shouldn't, but....). I'll add a follow-up item for this 👍

iamgabrielma replied:
https://github.com/woocommerce/woocommerce-ios/issues/13679

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleEntryPointView.swift:
```diff
@@ -9,14 +9,14 @@ struct PointOfSaleEntryPointView: View {
     @StateObject private var itemListViewModel: ItemListViewModel
 
-    private let hideAppTabBar: ((Bool) -> Void)
+    private let isPointOfSaleModeEnabled: ((Bool) -> Void)
```
```suggestion
    private let onPointOfSaleModeActiveStateChange: ((Bool) -> Void)
```

iamgabrielma replied:
Updated: https://github.com/woocommerce/woocommerce-ios/pull/13670/commits/d75f81f00b281dc596d6b725bcb8d12b7d99c910

joshheald reviewed (approved):
Thanks, this works well.

I've made some naming and style suggestions – primarily, "enabled" suggests that POS is not turned off, or prevented from working. "Active" suggests it's in-use, which is more accurate for this code.

I tested:
 - In the main app, after first launch without enabling POS mode, in-app notifications are always shown
 - In POS mode, no in-app notices are shown.
 - In the main app, in-app notices are shown for app and web orders.
 - When backgrounding the app from POS mode, background notifications are received as normal.
 - When foregrounding the app back to POS mode again, In-App notifications are not shown.
 - When moving to the main app via the `+ Create order in store management` link from the info panel, in-app notifications are shown as expected.

iamgabrielma committed:
Rename variables for clarity

iamgabrielma commented:
Thanks for the review!

> t's not even that bad in POS with the phone in portrait. Well, it is, but it feels like there's a route through. I wonder if we had the product list and cart stacked vertically, with horizontal scrolling for each, how that would feel... maybe weird, but weird can be good...

I can definitely see some sort of vertical stack setup with horizontal scrolling for POS on-the-go that could work for certain use cases, like restaurants where the operator goes to the table to take the order. Looks kind of nice with the big buttons and minimal info on iPhone.
